### PR TITLE
Changed MeshAssimp to use snorm16 if uv values are in the range [-1, 1]

### PR DESCRIPTION
**Pull request by magicwhale**

### CODE REVIEW

File: samples/app/MeshAssimp.cpp [hunk 1]
```diff
@@ -63,6 +63,7 @@ using namespace math;
 using namespace utils;
 
 
+
```
Comment (prideout) [inline samples/app/MeshAssimp.cpp]: I have a pet peeve about lots of consecutive vertical whitespace :)

File: samples/app/MeshAssimp.cpp [hunk 2]
```diff
@@ -213,6 +214,41 @@ Texture* MeshAssimp::createOneByOneTexture(uint32_t pixel) {
...
+            continue;
+        }
+        const float3* uv = reinterpret_cast<const float3*>(mesh->mTextureCoords[uvIndex]);
+        const size_t numVertices = mesh->mNumVertices;
+        const size_t numFaces = mesh->mNumFaces;
+        if (numVertices == 0 || numFaces == 0) {
+            continue;
+        }
+        for (size_t j = 0; j < numVertices; j++) {
+            if (uv) {
```
Comment (prideout) [inline samples/app/MeshAssimp.cpp]: Hoist up this `if`

File: samples/app/MeshAssimp.cpp [hunk 3]
```diff
@@ -703,40 +851,40 @@ bool MeshAssimp::setFromFile(const Path& file, std::vector<uint32_t>& outIndices
...
+                            asset.indices.push_back(uint32_t(face.mIndices[k] + indicesOffset));
                         }
                     }
 
                     uint32_t materialId = mesh->mMaterialIndex;
                     aiMaterial const* material = scene->mMaterials[materialId];
 
                     aiString name;
-                    std::string materialName;
+                    std::__1::string materialName;
```
Comment (prideout) [inline samples/app/MeshAssimp.cpp]: __1?

File: tools/filamesh/src/main.cpp [hunk 4]
```diff
@@ -322,8 +322,8 @@ int main(int argc, char* argv[]) {
 
     // Check for acceptable assimp data and determine UV bounds.
     preprocessNode(scene, node);
-    g_snormUVs = g_minUV.x >= 0.0f && g_minUV.x <= 1.0f && g_maxUV.x >= 0.0f && g_maxUV.x <= 1.0f &&
-                 g_minUV.y >= 0.0f && g_minUV.y <= 1.0f && g_maxUV.y >= 0.0f && g_maxUV.y <= 1.0f;
+    g_snormUVs = g_minUV.x >= -1.0f && g_minUV.x <= 1.0f && g_maxUV.x >= -1.0f && g_maxUV.x <= 1.0f &&
+                 g_minUV.y >= -1.0f && g_minUV.y <= 1.0f && g_maxUV.y >= -1.0f && g_maxUV.y <= 1.0f;
```
Comment (prideout) [inline tools/filamesh/src/main.cpp]: Thanks for fixing this.

File: samples/app/MeshAssimp.cpp [hunk 5]
```diff
@@ -213,6 +214,41 @@ Texture* MeshAssimp::createOneByOneTexture(uint32_t pixel) {
     return texturePtr;
 }
 
+void getMinMaxUV(const aiScene *scene, const aiNode* node, float2 &minUV, float2 &maxUV, int uvIndex) {
+    for (size_t i = 0; i < node->mNumMeshes; ++i) {
+        const aiMesh* mesh = scene->mMeshes[node->mMeshes[i]];
+        if (!mesh->HasNormals() || !mesh->HasTextureCoords(uvIndex)) {
```
Comment (bejado) [inline samples/app/MeshAssimp.cpp]: Why do you check `HasNormals` here?

File: samples/app/MeshAssimp.cpp [hunk 6]
```diff
@@ -213,6 +214,41 @@ Texture* MeshAssimp::createOneByOneTexture(uint32_t pixel) {
...
+            continue;
+        }
+        for (size_t j = 0; j < numVertices; j++) {
+            if (uv) {
+                minUV = min(uv[j].xy, minUV);
+                maxUV = max(uv[j].xy, maxUV);
+            }
+        }
+    }
+    for (size_t i=0 ; i<node->mNumChildren ; ++i) {
```
Comment (bejado) [inline samples/app/MeshAssimp.cpp]:
```suggestion
    for (size_t i = 0; i < node->mNumChildren; ++i) {
```

File: samples/app/MeshAssimp.cpp [hunk 7]
```diff
@@ -647,21 +688,128 @@ bool MeshAssimp::setFromFile(const Path& file, std::vector<uint32_t>& outIndices
...
-        mat4f parentTransform = parentIndex >= 0 ? outMeshes[parentIndex].accTransform : mat4f();
-        outMeshes.back().accTransform = parentTransform * current;
+        asset.positions.reserve(asset.positions.size() + totalVertexCount);
+        asset.tangents.reserve(asset.tangents.size() + totalVertexCount);
+        asset.texCoords0.reserve(asset.texCoords0.size() + totalVertexCount);
+        asset.texCoords1.reserve(asset.texCoords1.size() + totalVertexCount);
+        asset.indices.reserve(asset.indices.size() + totalIndexCount);
+
+        float2 minUV0 = float2(std::numeric_limits<float>::max());
+        float2 maxUV0 = float2(-std::numeric_limits<float>::max());
```
Comment (bejado) [inline samples/app/MeshAssimp.cpp]: It's probably best to use `std::numeric_limits<float>::min()`

File: samples/app/MeshAssimp.cpp [hunk 8]
```diff
@@ -647,21 +688,128 @@ bool MeshAssimp::setFromFile(const Path& file, std::vector<uint32_t>& outIndices
...
+            }
+        }
+        return true;
+    }
+    return false;
+}
+
+template<bool SNORMUV0, bool SNORMUV1>
+void MeshAssimp::processNode(Asset& asset,
+                             std::map<std::string,
```
Comment (bejado) [inline samples/app/MeshAssimp.cpp]: See the style guide on formatting arguments.

File: samples/app/MeshAssimp.cpp [hunk 9]
```diff
@@ -787,95 +935,27 @@ bool MeshAssimp::setFromFile(const Path& file, std::vector<uint32_t>& outIndices
...
+                    asset.meshes.back().parts.push_back({
                             indexBufferOffset, indicesCount, materialName,
                             baseColor, opacity, metallic, roughness, reflectance
                     });
                 }
             }
         }
-        if (node->mNumMeshes > 0) {
-            outMeshes.back().count = totalIndices;
+    if (node->mNumMeshes > 0) {
```
Comment (bejado) [inline samples/app/MeshAssimp.cpp]: Indentation issues here

File: samples/app/MeshAssimp.cpp (hunk 7 above)
Comment (prideout) [reply samples/app/MeshAssimp.cpp]: Actually `min` is wrong, `lowest` is correct.  Michael you can change the filamesh code to use `lowest` too.

File: samples/app/MeshAssimp.cpp (hunk 5 above)
Comment (prideout) [reply samples/app/MeshAssimp.cpp]: Good catch, this is a check that got copied from filamesh, not required here

File: samples/app/MeshAssimp.cpp [hunk 10]
```diff
@@ -647,21 +686,128 @@ bool MeshAssimp::setFromFile(const Path& file, std::vector<uint32_t>& outIndices
...
-        mat4f parentTransform = parentIndex >= 0 ? outMeshes[parentIndex].accTransform : mat4f();
-        outMeshes.back().accTransform = parentTransform * current;
+        asset.positions.reserve(asset.positions.size() + totalVertexCount);
+        asset.tangents.reserve(asset.tangents.size() + totalVertexCount);
+        asset.texCoords0.reserve(asset.texCoords0.size() + totalVertexCount);
+        asset.texCoords1.reserve(asset.texCoords1.size() + totalVertexCount);
+        asset.indices.reserve(asset.indices.size() + totalIndexCount);
+
+        float2 minUV0 = float2(std::numeric_limits<float>::max());
+        float2 maxUV0 = float2(std::numeric_limits<float>::min());
```
Comment (prideout) [inline samples/app/MeshAssimp.cpp]: `numeric_limits::min` is a minimum positive value, you should use `numeric_limits::lower` here.

File: samples/app/MeshAssimp.cpp [hunk 11]
```diff
@@ -647,21 +686,128 @@ bool MeshAssimp::setFromFile(const Path& file, std::vector<uint32_t>& outIndices
...
+        getMinMaxUV(scene, node, minUV0, maxUV0, 0);
+        float2 minUV1 = float2(std::numeric_limits<float>::max());
+        float2 maxUV1 = float2(-std::numeric_limits<float>::max());
```
Comment (prideout) [inline samples/app/MeshAssimp.cpp]: `std::numeric_limits<float>::lower()`

File: samples/app/MeshAssimp.cpp [hunk 12]
```diff
@@ -647,21 +686,128 @@ bool MeshAssimp::setFromFile(const Path& file, std::vector<uint32_t>& outIndices
...
+        getMinMaxUV(scene, node, minUV1, maxUV1, 1);
+
+        asset.snormUV0 = minUV0.x >= -1.0f && minUV0.x <= 1.0f && maxUV0.x >= -1.0f && maxUV0.x <= 1.0f &&
+                     minUV0.y >= -1.0f && minUV0.y <= 1.0f && maxUV0.y >= -1.0f && maxUV0.y <= 1.0f;
+        
+        asset.snormUV1 = minUV1.x >= -1.0f && minUV1.x <= 1.0f && maxUV1.x >= -1.0f && maxUV1.x <= 1.0f &&
+                         minUV1.y >= -1.0f && minUV1.y <= 1.0f && maxUV1.y >= -1.0f && maxUV1.y <= 1.0f;
```
Comment (prideout) [inline samples/app/MeshAssimp.cpp]: inconsistent indentation

File: samples/app/MeshAssimp.cpp [hunk 13]
```diff
@@ -647,21 +686,128 @@ bool MeshAssimp::setFromFile(const Path& file, std::vector<uint32_t>& outIndices
...
+                                        scene, isGLTF, deep, matCount, node, -1, depth);
+            } else {
+                processNode<false, false>(asset, outMaterials,
+                                        scene, isGLTF, deep, matCount, node, -1, depth);
+            }
+        }
 
-        for (size_t i = 0; i < node->mNumMeshes; i++) {
+
+        std::cout << "Hierarchy depth = " << depth << std::endl;
```
Comment (prideout) [inline samples/app/MeshAssimp.cpp]: Remove

**Pull request merged**